tests/e2e/upgrade_test.go

```diff
@@ -0,0 +1,140 @@
+		Expect(err).To(Not(HaveOccurred()))
+
+		// Get Fleet Version before-upgrade
+		getFleetImageVersion := []string{"get", "pod",
+		"--namespace", "cattle-fleet-system",
+		"-l", "app=fleet-controller",
+		"-o", "jsonpath={.items[*].status.containerStatuses[*].image}",
+		}
+
+		// Execute the shell command to get version before upgrade
+		fleetVersionBeforeUpgrade, err := kubectl.RunWithoutErr(getFleetImageVersion...)
+
+		// Upgrade Rancher Manager
+		// NOTE: Don't check the status, we can have false-positive here...
+		//       Better to check the rollout after the upgrade, it will fail if the upgrade failed
+		_ = rancher.DeployRancherManager(
+			rancherHostname,
+			rancherUpgradeChannel,
+			rancherUpgradeVersion,
+			rancherUpgradeHeadVersion,
+			"None",
+			"None",
+		)
+
+		// Wait for Rancher Manager to be restarted
+		// NOTE: 1st or 2nd rollout command can sporadically fail, so better to use Eventually here
+		Eventually(func() string {
+			status, _ := kubectl.RunWithoutErr(
+				"rollout",
+				"--namespace", "cattle-system",
+				"status", "deployment/rancher",
+			)
+			return status
+		}, tools.SetTimeout(4*time.Minute), 30*time.Second).Should(ContainSubstring("successfully rolled out"))
+
+		// Check that all Rancher Manager pods are running
+		Eventually(func() error {
+			checkList := [][]string{
+				{"cattle-system", "app=rancher"},
+				{"cattle-fleet-local-system", "app=fleet-agent"},
+				{"cattle-system", "app=rancher-webhook"},
+			}
+			return rancher.CheckPod(k, checkList)
+		}, tools.SetTimeout(3*time.Minute), 10*time.Second).Should(Not(HaveOccurred()))
+
+		// Check that all pods are using the same version
+		Eventually(func() int {
+			out, _ := kubectl.RunWithoutErr(getImageVersion...)
+			return len(strings.Fields(out))
+		}, tools.SetTimeout(3*time.Minute), 10*time.Second).Should(Equal(1))
+
+		// Get after-upgrade Rancher Manager version
+		// and check that it's different to the before-upgrade version
+		versionAfterUpgrade, err := kubectl.RunWithoutErr(getImageVersion...)
+		Expect(err).To(Not(HaveOccurred()))
+		Expect(versionAfterUpgrade).To(Not(Equal(versionBeforeUpgrade)))
+
+		// Function to check if all Fleet pods are updated and running the new version
+		isFleetControllerUpgradeComplete := func() bool {
+			// Check the rollout status of Fleet pods to ensure they are updated
+			rolloutStatus, err := kubectl.RunWithoutErr(
+				"rollout",
+				"--namespace", "cattle-fleet-system",
+				"status", "deployment/fleet-controller",
+			)
+			if err != nil {
+					return false
+			}
+
+			// Check if the rollout has completed successfully
+			return strings.Contains(rolloutStatus, `deployment "fleet-controller" successfully rolled out`)
+		}
+
+		// Wait for the upgrade to complete by checking if the Fleet rollout is complete
+		Eventually(isFleetControllerUpgradeComplete, tools.SetTimeout(10*time.Minute), 20*time.Second).Should(BeTrue())
```

Comment:
If this logic is not reused, it could be a simple `Eventually` block, eg.:
```
Eventually(func() string{
    // ....
    return rolloutStatus
}, 10 * time.Minute, 20 * time.Second).Should(ContainSubstring(`deployment "fleet-controller" successfully rolled out`))
```